# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: uriloader/exthandler/nsExternalHelperAppService.h
# Commit: baffa8a7b961
# Full Hash: baffa8a7b9618e9a02e4bd808adb71dead99cce5
# Author: Matt Woodrow <mwoodrow@mozilla.com>
# Date: 2019-11-12 09:43:07
# Regressor Bug: 1589270
# File Overlap Count: 1
# Description:
#   Bug 1589270 - Part 3: Convert nsExternalHelperApp to use BrowsingContext instead of nsIInterfaceRequestor. r=bzbarsky
#   
#   This also converts MaybeCloseWindowHelper, and results in the window close operations being always run in the parent (even without DocumentChannel).
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D49528
# ==============================================================================

diff -r 68bc2ae9b89a -r baffa8a7b961 uriloader/exthandler/nsExternalHelperAppService.h
--- a/uriloader/exthandler/nsExternalHelperAppService.h	Fri Nov 08 03:01:03 2019 +0000
+++ b/uriloader/exthandler/nsExternalHelperAppService.h	Fri Nov 08 04:35:05 2019 +0000
@@ -172,12 +172,11 @@
   nsCOMArray<nsIFile> mTemporaryPrivateFilesList;
 
  private:
-  nsresult DoContentContentProcessHelper(const nsACString& aMimeContentType,
-                                         nsIRequest* aRequest,
-                                         nsIInterfaceRequestor* aContentContext,
-                                         bool aForceSave,
-                                         nsIInterfaceRequestor* aWindowContext,
-                                         nsIStreamListener** aStreamListener);
+  nsresult DoContentContentProcessHelper(
+      const nsACString& aMimeContentType, nsIRequest* aRequest,
+      mozilla::dom::BrowsingContext* aContentContext, bool aForceSave,
+      nsIInterfaceRequestor* aWindowContext,
+      nsIStreamListener** aStreamListener);
 };
 
 /**
@@ -217,7 +216,7 @@
    * indicating why the request is handled by a helper app.
    */
   nsExternalAppHandler(nsIMIMEInfo* aMIMEInfo, const nsACString& aFileExtension,
-                       nsIInterfaceRequestor* aContentContext,
+                       mozilla::dom::BrowsingContext* aBrowsingContext,
                        nsIInterfaceRequestor* aWindowContext,
                        nsExternalHelperAppService* aExtProtSvc,
                        const nsAString& aFilename, uint32_t aReason,
@@ -233,17 +232,7 @@
    */
   void MaybeApplyDecodingForExtension(nsIRequest* request);
 
-  /**
-   * Get the dialog parent. Public for ExternalHelperAppChild::OnStartRequest.
-   */
-  nsIInterfaceRequestor* GetDialogParent() {
-    return mWindowContext ? mWindowContext : mContentContext;
-  }
-
-  void SetContentContext(nsIInterfaceRequestor* context) {
-    MOZ_ASSERT(!mWindowContext);
-    mContentContext = context;
-  }
+  void SetShouldCloseWindow() { mShouldCloseWindow = true; }
 
  protected:
   ~nsExternalAppHandler();
@@ -259,9 +248,9 @@
   nsCOMPtr<nsIMIMEInfo> mMimeInfo;
 
   /**
-   * The dom window associated with this request to handle content.
+   * The BrowsingContext associated with this request to handle content.
    */
-  nsCOMPtr<nsIInterfaceRequestor> mContentContext;
+  RefPtr<mozilla::dom::BrowsingContext> mBrowsingContext;
 
   /**
    * If set, the parent window helper app dialogs and file pickers
@@ -303,6 +292,12 @@
   bool mIsFileChannel;
 
   /**
+   * True if the ExternalHelperAppChild told us that we should close the window
+   * if we handle the content as a download.
+   */
+  bool mShouldCloseWindow;
+
+  /**
    * One of the REASON_ constants from nsIHelperAppLauncherDialog. Indicates the
    * reason the dialog was shown (unknown content type, server requested it,
    * etc).
@@ -350,6 +345,11 @@
    */
   nsCOMPtr<nsIArray> mRedirects;
   /**
+   * Get the dialog parent: the parent window that we can attach
+   * a dialog to when prompting the user for a download.
+   */
+  already_AddRefed<nsIInterfaceRequestor> GetDialogParent();
+  /**
    * Creates the temporary file for the download and an output stream for it.
    * Upon successful return, both mTempFile and mSaver will be valid.
    */